Add autosectionlabel_full_reference configuration variable

andrewjmaguire opened the pull request:
Subject: Enable the autosectionlabel extension to create more unique references 

### Feature or Bugfix
- Feature

### Purpose
- In a large document with many same-named sub-sections, the autosectionlabel extension will not create useful references that can be used to reference these non-uniquely named sub-sections.
This pull-request includes an option that makes the usable references that include the section name hierarchy in the reference string.

### Detail
- Add autosectionlabel_full_reference configuration variable


### Relates
- https://github.com/sphinx-doc/sphinx/issues/13075

Commit:
Add autosectionlabel_full_reference configuration variable

Commit:
python 3.10

Commit:
fix lint errors

picnixz commented on doc/usage/extensions/autosectionlabel.rst:
```diff
@@ -25,7 +25,9 @@ Internally, this extension generates the labels for each section.  If same
-unique.
+unique. Also the ``autosectionlabel_full_reference`` configuration variable can
+help ensure the generated refereences are unique across a document with similar
+section names at different levels.
```
```suggestion
unique. 

Use the :conf:`autosectionlabel_full_reference` configuration variable
to guarantee that the generated references are unique across a document
with similar section names at different levels.
```

picnixz commented on doc/usage/extensions/autosectionlabel.rst:
```diff
@@ -50,6 +52,24 @@ Configuration
+      -------
+
+      Sub Section
+      ~~~~~~~~~~~
```
```suggestion
   a colon. For instance:
   
   .. code-block:: rst

      Title
      =====

      Section
      -------

      Sub Section
      ~~~~~~~~~~~
```

picnixz commented on doc/usage/extensions/autosectionlabel.rst:
```diff
@@ -50,6 +52,24 @@ Configuration
+      ~~~~~~~~~~~
+
+   Then the reference of the third level section is
+   ``index:Title:Section:Sub Section``.
```
```suggestion
   The reference of the third level section ("Sub Section") will be
   ``index:Title:Section:Sub Section``.
```

picnixz commented on sphinx/ext/autosectionlabel.py:
```diff
@@ -29,6 +29,17 @@ def get_node_depth(node: Node) -> int:
     return i
 
 
+def get_node_parents(node: Node) -> list[str]:
```
```suggestion
def _find_parent_nodes(node: Node) -> list[str]:
```

andrewjmaguire replied:
I followed the same naming convention as used in the get_node_depth function above.
Do you want me to change that one as well to follow the same convention, i.e. rename it to _get_node_depth ?

picnixz replied:
I see where you're coming from. I'd prefer not bothering with keeping the naming convention. Historically, private and public APIs were named similarly, just that private ones were not documented. However, for new code, we're trying to avoid having public names if we can, especially for helper functions, so either `_get_parent_nodes` or `_find_parent_nodes` would be correct. 

- `_get_parent_nodes` would return a list
- `_find_parent_nodes` would yield elements one by one

Instead of constructing the list in the function itself, you can yield nodes one by one, hence the naming suggestion.

picnixz commented on sphinx/ext/autosectionlabel.py:
```diff
@@ -40,7 +51,15 @@ def register_sections_as_label(app: Sphinx, document: Node) -> None:
         if app.config.autosectionlabel_prefix_document:
-            name = nodes.fully_normalize_name(docname + ':' + ref_name)
+            if app.config.autosectionlabel_full_reference:
+                id_array = get_node_parents(node)
```
```suggestion
                id_array = _find_parent_nodes(node)
```

picnixz commented on sphinx/ext/autosectionlabel.py:
```diff
@@ -40,7 +51,15 @@ def register_sections_as_label(app: Sphinx, document: Node) -> None:
+                id_array = get_node_parents(node)
+                id_array.append(docname)
+                id_array.reverse()
+                id_array.append(ref_name)
```
```suggestion
				id_array = [docname, *reversed(id_array), refname]
```

picnixz commented on sphinx/ext/autosectionlabel.py:
```diff
@@ -40,7 +51,15 @@ def register_sections_as_label(app: Sphinx, document: Node) -> None:
+                name = nodes.fully_normalize_name(':'.join(id_array))
+                labelid = make_id(app.env, node.document, '', '.'.join(id_array))
+            else:
+                name = nodes.fully_normalize_name(docname + ':' + ref_name)
```
```suggestion
                name = nodes.fully_normalize_name(f'{docname}:{ref_name}')
```

picnixz commented on sphinx/ext/autosectionlabel.py:
```diff
@@ -60,6 +79,7 @@ def register_sections_as_label(app: Sphinx, document: Node) -> None:
 def setup(app: Sphinx) -> ExtensionMetadata:
     app.add_config_value('autosectionlabel_prefix_document', False, 'env')
     app.add_config_value('autosectionlabel_maxdepth', None, 'env')
+    app.add_config_value('autosectionlabel_full_reference', False, 'env')
```
```suggestion
    app.add_config_value('autosectionlabel_full_reference', False, 'env', bool)
```

andrewjmaguire replied:
Does this mean that bool needs to be added to this config variable too? i.e.
 app.add_config_value('autosectionlabel_prefix_document', False, 'env', bool)

picnixz replied:
Yes! That's how GitHub suggestions works (namely, you see the diff of my suggestion with your existing code). If you're wondering what `bool` means in this case, this simply means that we expect `autosectionlabel_prefix_document` to be of `bool` type.

picnixz commented on tests/roots/test-ext-autosectionlabel-full-reference/index.rst:
```diff
@@ -0,0 +1,48 @@
+* :ref:`index:Installation:For UNIX users:Linux:Command`
+* :ref:`index:Installation:For UNIX users:FreeBSD`
+* :ref:`index:Installation:For UNIX users:FreeBSD:Command`
+* :ref:`index:Installation:This one's got an apostrophe`
```
Can you check check with numbers in titles as well as with `:math:` blocks please? and with literal (double backquotes). Can we also use `:ref:` in a title? (if not, no need for the coverage).

Additionally, can you test when you specify a custom label before an inner title? as well as content in-between two titles (in other words, try to see if it's possible to construct a doctree with the following tree: "section -> non-section -> section") so that the parent of the second section is a non-section node.

andrewjmaguire replied:
Would adding an image in between suffice? e.g.
```

Section
-------

Here is an image or I could use a figure?

.. image:: test.png

Sub Section
~~~~~~~~~~~

See previous image...
```

picnixz replied:
Unfortunately, the image and subsection would be sibling nodes and not parent->child nodes. Maybe it's not possible to actually have a parent node that is not a section's node but I need to check that (it's midnight here so I won't do it today).

andrewjmaguire replied:
I have tested with numbers in sections

andrewjmaguire replied:
I have tried to added all the other comments, and added texinfo and latex tests and the test includes another rst file, to check how references worked there.
How is it looking now?

picnixz commented on tests/test_extensions/test_ext_autosectionlabel.py:
```diff
@@ -98,3 +98,78 @@ def test_autosectionlabel_maxdepth(app):
+
+
+@pytest.mark.sphinx('html', testroot='ext-autosectionlabel-full-reference')
+def test_autosectionlabel_full_reference(app):
```
Can you also add tests for the LaTeX output and texinfo/man if the extension is properly supported? (I think LaTeX should be fine but I'm not sure for texinfo/man).

picnixz commented on sphinx/ext/autosectionlabel.py:
```diff
@@ -29,6 +29,17 @@ def get_node_depth(node: Node) -> int:
 
+def get_node_parents(node: Node) -> list[str]:
+    parents = []
+    cur_node = node
+    while cur_node.parent != node.document:
+        cur_node = cur_node.parent
+        title = cast(nodes.title, cur_node[0])
+        ref_name = getattr(title, 'rawsource', title.astext())
+        parents.append(ref_name)
+    return parents
```
This will not be sufficient. You should walk the parent nodes that are only `nodes.section` and not other kind of nodes that could appear.

andrewjmaguire replied:
I am very new to Sphinx, can you give me an example, please?

picnixz replied:
Sure! I'll try to see if what I had in mind works tomorrow. If not, your algorithm would be correct (and we could also add `assert isinstance(node.parent, nodes.section)`)

picnixz reviewed (commented):
Thanks, that's a nice addition but I'd like you to address the following points (especially for improving test coverage because I think the current algorithm will not work if there are non-section nodes between two sections).

In addition, we'll need a CHANGELOG entry for this one since it's a new feature.

Commit:
Added autosectionlabel_full_reference to CHANGES.rst

In usage, updated as suggested in review
Added versionadded 8.2 for the new feature

Walked node tree looking only for nodes.section type.
Prefixed internally used functions with _.

Added bool type to add_config_value calls
Used Python f'' string for efficiency

Overrides document.settings.id_prefix so that suitable references
are created in case of any duplicate section hierarchy. Tests for these
are created too.

The autosectionlabel_full_reference tests now include another file
to check multiple file references.

When using autosectionlabel_full_reference in texinfo,
the filename is not prefixed in references. Added texinfo tests.

When using autosectionlabel_full_reference in latex,
the filename is not prefixed in references. Added latex tests

All new tests pass
All type-checks pass

Commit:
Fixes to pass ruff format and flake8 style-lint

picnixz commented on doc/usage/extensions/autosectionlabel.rst:
```diff
@@ -25,9 +25,11 @@ Internally, this extension generates the labels for each section.  If same
-section names at different levels.
+unique.
+
+Use the :conf:`autosectionlabel_full_reference` configuration variable
```
```suggestion
Use the :confval:`autosectionlabel_full_reference` configuration variable
```

picnixz commented on CHANGES.rst:
```diff
@@ -15,6 +15,8 @@ Deprecated
 Features added
 --------------
 
+* #13076: Add autosectionlabel_full_reference configuration variable.
```
```suggestion
* #13076: Add the :confval:`autosectionlabel_full_reference`
  configuration variable to create more precise labels.
```

andrewjmaguire replied:
done

picnixz commented on sphinx/ext/autosectionlabel.py:
```diff
@@ -20,7 +20,7 @@
 
 
-def get_node_depth(node: Node) -> int:
+def _get_node_depth(node: Node) -> int:
```
Ah maybe I wasn't clear. Old APIs are fine to be kept untouched (because we don't want to break packages) but it's good if *new* APIs can be designed with a good scope.

picnixz commented on sphinx/ext/autosectionlabel.py:
```diff
@@ -29,43 +29,57 @@ def get_node_depth(node: Node) -> int:
 
 
-def get_node_parents(node: Node) -> list[str]:
+def _find_node_parents(node: Node) -> list[str]:
```
I would revert this change because it could be used in the wild (and to make the diff smaller.

andrewjmaguire replied:
Have renamed to _get_all_node_parent_section_titles to describe
more fully what the function does and _ prefix is just for internal use

picnixz reviewed (commented):
I have a lot of work to do so I'm only doing a half-review now (sorry). I'll try to come back later this week or the week after, though I will have to travel afterwards (sorry x2).

Commit:
Use :confval: to refer to autosectionlabel_full_reference in documentation
Undo _get_node_depth rename
Change _find_node_parents to _get_all_node_parent_section_titles to describe
more fully what the function does and _ prefix is just for internal use.

Commit:
Use getattr for accessing autosectionlabel_full_reference in latex and texinfo writers

Commit:
Revert "python 3.10"

This reverts commit 9072d6213b47b3a5e11f669a512ca2d64616a414.

Commit:
Use getattr for accessing autosectionlabel_full_reference in latex (missed one place)

Commit:
use one-liner for constructing id_array parts

Commit:
fix typo

Commit:
Add the generated full section label id to the list of node ids.
This makes the full section id available to be used for cross references.

Note: The original short id is still needed by other functionality.
I.e. the full id cannot replace the short one, `node['ids'][0] = labelid`
because then TOC and sectnum:: functionality is broken in the HTML output.

Commit:
Remove blank line
